burn-wgpu/src/compute/server.rs

```diff
@@ -209,29 +201,70 @@ where
 
         self.submit();
 
-        let buffer_slice = buffer_dest.slice(..);
+        BufferReader::new(buffer_dest)
+    }
+}
+
+#[derive(new)]
+struct BufferReader {
+    buffer: wgpu::Buffer,
+}
+
+impl BufferReader {
+    #[cfg(target_family = "wasm")]
+    async fn read(self, device: alloc::sync::Arc<wgpu::Device>) -> Vec<u8> {
+        self.read_async(&device).await
+    }
+
+    #[cfg(not(target_family = "wasm"))]
+    fn read(self, device: &wgpu::Device) -> Vec<u8> {
+        pollster::block_on(self.read_async(device))
+    }
+
+    async fn read_async(&self, device: &wgpu::Device) -> Vec<u8> {
+        let buffer_slice = self.buffer.slice(..);
         let (sender, receiver) = futures_intrusive::channel::shared::oneshot_channel();
         buffer_slice.map_async(wgpu::MapMode::Read, move |v| {
             sender
                 .send(v)
                 .expect("Unable to send buffer slice result to async channel.")
         });
 
-        self.device.poll(wgpu::Maintain::Wait);
+        device.poll(wgpu::Maintain::Wait);
 
-        let result = pollster::block_on(receiver.receive());
+        let result = receiver.receive().await;
 
         if let Some(Ok(())) = result {
             let data = buffer_slice.get_mapped_range();
             let result = bytemuck::cast_slice(&data).to_vec();
 
             drop(data);
-            buffer_dest.unmap();
+            self.buffer.unmap();
             result
         } else {
             panic!("Unable to read buffer {:?}", result)
         }
     }
+}
+
+impl<MM> ComputeServer for WgpuServer<MM>
+where
+    MM: MemoryManagement<WgpuStorage>,
+{
+    type Kernel = Box<dyn Kernel>;
+    type Storage = WgpuStorage;
+    type MemoryManagement = MM;
+
+    fn read(&mut self, handle: &server::Handle<Self>) -> Reader<Vec<u8>> {
+        #[cfg(target_family = "wasm")]
+        {
+            let future = self.buffer_reader(handle).read(self.device.clone());
+            return Reader::Future(Box::pin(future));
+        }
+
+        #[cfg(not(target_family = "wasm"))]
+        Reader::Concrete(self.buffer_reader(handle).read(&self.device))
+    }
```

Comment:
It might be worth taking this discussion outside of the closed PR because I think we have something important to resolve in regards to async/sync data read and people might want to see it. I am not sure if we have an issue or thread regarding this.

Reply:
@AlexErrant reading data is by nature a blocking operation, so no there isn't any problem doing busy waiting on a thread where its only job is to run kernels. We could update the file `burn-compute/src/channel/mpsc.rs` to support running the server on a web worker thread, and change the signature to returns `Vec<u8>` instead of `Reader<Vev<u8>>` from the client side.

```rust
/// Create a new mpsc compute channel.
    pub fn new(mut server: Server, bound: usize) -> Self {
        let (sender, receiver) = mpsc::sync_channel(bound);

        let _handle = web_workers::spawn(move || async {
            while let Ok(message) = receiver.recv() {
                match message {
                    Message::Read(handle, callback) => {
                        let data = server.read(&handle);
                        core::mem::drop(handle);
                        let data = data.await; // or something similar
                        callback.send(data).unwrap();
                    }
                   _ => todo!("same as before"),
                };
            }
        });
```